openedx/core/djangoapps/content/search/api.py

```diff
@@ -0,0 +1,370 @@
+"""
+Content index and search API using Meilisearch
+"""
+from __future__ import annotations
+
+import logging
+import time
+from contextlib import contextmanager
+from datetime import datetime, timedelta, timezone
+from typing import Callable, Generator
+
+import meilisearch
+from django.conf import settings
+from django.contrib.auth import get_user_model
+from django.core.cache import cache
+from meilisearch.errors import MeilisearchError
+from meilisearch.models.task import TaskInfo
+from opaque_keys.edx.keys import UsageKey
+
+from openedx.core.djangoapps.content.search.documents import (
+    STUDIO_INDEX_NAME,
+    Fields,
+    meili_id_from_opaque_key,
+    searchable_doc_for_course_block,
+    searchable_doc_for_library_block
+)
+from openedx.core.djangoapps.content_libraries import api as lib_api
+from xmodule.modulestore import ModuleStoreEnum
+from xmodule.modulestore.django import modulestore
+
+from .documents import Fields, searchable_doc_for_course_block, searchable_doc_for_library_block
+
+log = logging.getLogger(__name__)
+
+User = get_user_model()
+
+STUDIO_INDEX_NAME = "studio_content"
+INDEX_NAME = settings.MEILISEARCH_INDEX_PREFIX + STUDIO_INDEX_NAME
+
+_MEILI_CLIENT = None
+_MEILI_API_KEY_UID = None
+
+LOCK_EXPIRE = 5 * 60  # Lock expires in 5 minutes
+
+
+@contextmanager
+def _index_rebuild_lock(index_name: str) -> Generator[str, None, None]:
+    """
+    Lock to prevent that more than one rebuild is running at the same time
+    """
+    timeout_at = time.monotonic() + LOCK_EXPIRE
+    lock_id = f"lock-meilisearch-index-{index_name}"
+    new_index_name = index_name + "_new"
+
+    while True:
+        status = cache.add(lock_id, new_index_name, LOCK_EXPIRE)
+        if status:
+            # Lock acquired
+            try:
+                yield new_index_name
+            finally:
+                # Release the lock
+                cache.delete(lock_id)
+                return
+
+        if time.monotonic() > timeout_at:
+            raise TimeoutError("Timeout acquiring lock")
+
+        time.sleep(1)
+
+
+def _get_running_rebuild_index_name(index_name: str) -> str | None:
+    lock_id = f"lock-meilisearch-index-{index_name}"
+
+    return cache.get(lock_id)
+
+
+def _get_meilisearch_client():
+    """
+    Get the Meiliesearch client
+    """
+    global _MEILI_CLIENT  # pylint: disable=global-statement
+
+    if _MEILI_CLIENT is not None:
+        return _MEILI_CLIENT
+
+    # Connect to Meilisearch
+    if not settings.MEILISEARCH_ENABLED:
+        raise RuntimeError("MEILISEARCH_ENABLED is not set - search functionality disabled.")
```

Comment:
Should this check be at the top so it raises the error instead of returning the client when its not enabled:

```suggestion
    # Connect to Meilisearch
    if not settings.MEILISEARCH_ENABLED:
        raise RuntimeError("MEILISEARCH_ENABLED is not set - search functionality disabled.")
    
    if _MEILI_CLIENT is not None:
        return _MEILI_CLIENT
```